core/src/typecheck/mod.rs

```diff
@@ -99,34 +101,228 @@ pub type Environment = GenericEnvironment<Ident, UnifType>;
 
 /// Mapping from wildcard ID to inferred type
 pub type Wildcards = Vec<Types>;
+
 /// Unification variable or type constants unique identifier.
 pub type VarId = usize;
 
+/// Variable levels. Levels are used in order to implement higher-rank polymorphism in a sound way:
+/// we need to associate to each unification variable and rigid type variable a level, which
+/// depends on when those variables were introduced, and to forbid some unifications if a condition
+/// on levels is not met.
+#[derive(Clone, Copy, Ord, Eq, PartialEq, PartialOrd, Debug)]
+pub struct VarLevel(NonZeroU16);
```

Comment:
I believe it was mix of:
- although there's no gain on one `var_level`, I expect gain when var_levels are packed up in a struct. Together with the `NonZero` optimization, you can expect two optionals var levels to be packed in 32bits. Indeed the size of `VarLevelsData` is currently 4 bytes. Granted, in this case it doesn't win much, because I don't think `VarLevelsData` is packed with anything that isn't 64 bits aligned elsewhere.
- given the previous point, I just expect the level to never ever reach anything above a few hundreds, so I thought `u16` was a safe bet. I guess we can still go `u32` and keep `VarLevelsData` packed up within 64 bits.

Reply:
I suspect I'll need to pack an additional boolean in `VarLevelsData` for cycle detection/delayed occurs check. Let's keep it `u16` for now and see later if we can extend it without changing the memory representation